packages/teuchos/comm/test/Comm/teuchosSubcommTest.cpp

```diff
@@ -0,0 +1,42 @@
+// Test to exercise fix to Teuchos::Comm's createSubcommunicator, which
+// had leaked memory.
+// The fix added MPI_Comm_free to the opaqueWrappers. 
+// 8/2018  This test hangs of platform waterman.
+
+#include <stdio.h>
+#include <mpi.h>
+#include "Teuchos_CommHelpers.hpp"
+#include "Teuchos_DefaultComm.hpp"
+#include "Teuchos_RCP.hpp"
+#include "Teuchos_ArrayView.hpp"
+
+
+int main(int narg, char **arg)
+{
+  Teuchos::GlobalMPISession mpiSession(&narg,&arg);
+
+  Teuchos::RCP<const Teuchos::Comm<int> >
```

Comment:
`Teuchos::DefaultComm<int>::getComm()` does something more than just return a wrapped `MpiComm` -- it schedules the resulting object for deletion at `MPI_Finalize`.  See [the implementation here](https://github.com/trilinos/Trilinos/blob/b783a65a6faed6a40372d404080ba2834f6af155/packages/teuchos/comm/src/Teuchos_DefaultComm.hpp#L288).  This means that any use of the result of this function after `MPI_Finalize` has been called is user error.

This code example is correct in standard C++, because destructor order is supposed to be the reverse of constructor order.  (`GlobalMPISession::~GlobalMPISession()` calls `MPI_Finalize`.)  However, it could be that some compiler bug on this platform somehow has allowed these destructors to be called out of order.  That _should_ be OK, since this code does not use `comm` before end of scope.